### PR TITLE
Cr 120 respondent authenticated event

### PR DESCRIPTION
**Pull request by MartinHumphrey**
# Motivation and Context
On retrieval of UAC and Case details from the repository publish a Respondent Authenticated event to the RabbitMQ event exchange, prior to sending the response to the UAC authentication request.

# What has changed
UniqueAccessCodeServiceImpl changed to send Respondent Authenticated event. RespondentEventPublisher adapted to cater for sending Survey Launched or Respondent Authenticated events. New EventBuilder class concerned with creation of events introduced. RespondentHomeEndpoint, RespondentHomeServiceImpl changed to adapt to generic event message construction and publishing. 

# How to test?
All changes are supported by unit tests. 

To test manually run application and create a new queue in RabbitMQ bound to the events exchange with the "event.response.*" binding. 

In RabbitMQ publish the following data to the events exchange with the "event.uac.update" binding: 

{ 
     "event" : { 
           "type" : "UAC_UPDATED", 
           "source" : "CASE_SERVICE", 
           "channel" : "RM", 
           "dateTime" : "2011-08-12T20:17:46.384Z", 
           "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93" 
      }, 
      "payload" : {
              "uac" : { 
                  "uacHash": "8a9d5db4bbee34fd16e40aa2aaae52cfbdf1842559023614c30edb480ec252b4", 
                  "active" : true, 
                  "questionnaireId":"1110000009", 
                  "caseType":"HH", 
                  "region":"E", 
                  "caseId":"dc4477d1-dd3f-4c69-b181-7ff725dc9fa4", 
                  "collectionExerciseId":"a66de4dc-3c3b-11e9-b210-d663bd873d93" 
              } 
      } 
}

In RabbitMQ publish the following data to the events exchange with the "event.case.lifecycle" binding: 

{ 
     "event" : { 
           "type" : "CASE_UPDATED", 
           "source" : "CASE_SERVICE", 
           "channel" : "RM", 
           "dateTime" : "2011-08-12T20:17:46.384Z", 
           "transactionId" : "c45de4dc-3c3b-11e9-b210-d663bd873d93" 
     }, 
     "payload" : { 
             "collectionCase" : { 
                        "id":"dc4477d1-dd3f-4c69-b181-7ff725dc9fa4", 
                        "caseRef": "10000000010", 
                        "survey": "CENSUS", 
                        "collectionExerciseId": "a66de4dc-3c3b-11e9-b210-d663bd873d93", 
                        "address": { 
                                "addressLine1": "1 main street", 
                                "addressLine2": "upper upperingham", 
                                "addressLine3": "", 
                                "townName": "upton", 
                                "postcode": "UP103UP", 
                                "region": "E", 
                                "latitude": "50.863849", 
                                "longitude": "-1.229710", 
                                "uprn":"123456", 
                                "arid":"XXXXX", 
                                "addressType":"HH", 
                                "estabType":"XXX" 
                        }, 
                        "contact": { 
                                "title":"Ms", 
                                "forename":"jo", 
                                "surname":"smith", 
                                "email":"me@example.com", 
                                "telNo":"+447890000000" 
                        }, 
                        "state":"ACTIONABLE", 
                        "actionableFrom":"2011-08-12T20:17:46.384Z" 
             } 
     } 
}

Postman or Curl request http://localhost:8071/uacs/w4nwwpphjjptp7fn will get a UAC authentication response which the RH UI uses to authenticate and you should see a Respondent Authentication event in the queue you manually created above.

You can also test the SurveyLaunched endpoint which has been changed by sending a POST request to http://localhost:8071/surveyLaunched with a body such as:

{
	"questionnaireId": "23434234234",
	"caseId": "3fa85f64-5717-4562-b3fc-2c963f66afa6"
}

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    private String channel;
+
+    EventType(String source, String channel) {
+      this.source = source;
+      this.channel = channel;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends GenericEvent, S> T buildEvent(EventType type, S payload)
```
Comment (PeterBochel) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java]: This code would be simpler if it was implemented as 2 distinct methods. All the conditional logic and associated exception could then be removed. 
There are only 2 callers to the method, and both have fixed arguments, so the flexibility provided by taking generic arguments is not currently needed. YAGNI !

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java [hunk 2]
```diff
@@ -57,6 +62,7 @@ public UniqueAccessCodeDTO getUniqueAccessCodeData(String uac) throws CTPExcepti
         if (caseMatch.isPresent()) {
           caseMapperFacade.map(caseMatch.get(), data);
           data.setCaseStatus(CaseStatus.OK);
+          sendRespondentAuthenticatedEvent(data);
```
Comment (PeterBochel) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java]: Perhaps the method now needs needs to renamed, as it seems very surprising that a 'get' method has the side effect of sending a message.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java (hunk 1 above)
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java]: Could use overloaded methods in a Builder type implementation. Would remove the need for Generics and a Switch statement. The idea was as you suggest the buildEvent method could be easily extended in future for all events keeping the same logic each time without having to add a new method every time. As you say are we going to need it?

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java (hunk 2 above)
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/UniqueAccessCodeServiceImpl.java]: Renamed method as verbally agreed.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java (hunk 1 above)
Comment (MartinHumphrey) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java]: Simplified Eventbuilder as two distinct methods removing generics and associated exceptions as suggested.  Removed a lot of unnecessary complication.

**Pull request merged**